File: app/src/main/java/com/iug/coursehub/feature/dashboard/courses_screen/CoursesAdapter.java
package com.iug.coursehub.feature.dashboard.courses_screen;

import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.iug.coursehub.data.local.db.entity.Course;
import com.iug.coursehub.databinding.ItemCourseBinding;

import android.view.LayoutInflater;

import java.util.List;

public class CoursesAdapter extends RecyclerView.Adapter<CoursesAdapter.CourseViewHolder> {

    private List<Course> courses;
    private OnShowLessonClick onShowLessonsClick;
    private OnEditCourseClick onEditCourseClick;
    private OnDeleteCourseClick onDeleteCourseClick;

    public CoursesAdapter(
            List<Course> courses,
            OnShowLessonClick onShowLessonsClick,
            OnEditCourseClick onEditCourseClick,
            OnDeleteCourseClick onDeleteCourseClick
    ) {
        this.courses = courses;
        this.onShowLessonsClick = onShowLessonsClick;
        this.onEditCourseClick = onEditCourseClick;
        this.onDeleteCourseClick = onDeleteCourseClick;
    }

    @NonNull
    @Override
    public CourseViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new CourseViewHolder(
            ItemCourseBinding.inflate(
                LayoutInflater.from(parent.getContext()),
                parent,
                false
            )
        );
    }

    @Override
    public void onBindViewHolder(@NonNull CourseViewHolder holder, int position) {
        Course course = courses.get(position);
        holder.getBinding().tvCourseName.setText(course.getName());
        holder.getBinding().tvPrice.setText(course.getPrice() + "");

        holder.getBinding().btnShowLessons.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                onShowLessonsClick.onShowLessonsClick(course.getId());
            }
        });

        holder.getBinding().btnEdit.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                onEditCourseClick.onEditCourseClick(course.getId());
            }
        });

        holder.getBinding().btnDelete.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                onDeleteCourseClick.onDeleteCourseClick(course);
            }
        });
    }

    @Override
    public int getItemCount() {
        return courses.size();
    }

    static class CourseViewHolder extends RecyclerView.ViewHolder {

        private ItemCourseBinding binding;

        public ItemCourseBinding getBinding() {
            return binding;
        }

        public CourseViewHolder(ItemCourseBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }
    }
}

File: app/src/main/java/com/iug/coursehub/feature/dashboard/courses_screen/CoursesAdapter.kt
package com.iug.coursehub.feature.dashboard.courses_screen

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.iug.coursehub.data.local.db.entity.Course
import com.iug.coursehub.databinding.ItemCourseBinding

class CoursesAdapter(
    private var courses: List<Course>,
    private val onShowLessonsClick: (Int) -> Unit,
    private val onEditCourseClick: (Int) -> Unit,
    private val onDeleteCourseClick: (Course) -> Unit
) : RecyclerView.Adapter<CoursesAdapter.CourseViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CourseViewHolder {
        return CourseViewHolder(
            ItemCourseBinding.inflate(
                LayoutInflater.from(parent.context),
                parent,
                false
            )
        )
    }

    override fun onBindViewHolder(holder: CourseViewHolder, position: Int) {
        holder.binding.apply {
            val course = courses[position]
            tvCourseName.text = course.name
            tvPrice.text = "$${course.price}"

            btnShowLessons.setOnClickListener {
                onShowLessonsClick(course.id!!)
            }

            btnEdit.setOnClickListener {
                onEditCourseClick(course.id!!)
            }

            btnDelete.setOnClickListener {
                onDeleteCourseClick(course)
            }
        }
    }

    override fun getItemCount(): Int {
        return courses.size
    }


    class CourseViewHolder(val binding: ItemCourseBinding)
        : RecyclerView.ViewHolder(binding.root)
}